[BUG] Tidy command - send on closed channel v0.6.0

When executing the tidy command, not all messages are processed, and the process exits with `panic: send on closed channel`.

During development, this seemed to work reliably in a real use case, but failed during the integration tests. The integration steps passed successfully when a delay was added before closing the queue.

**To Reproduce**
Add 3 messages on the queue (e.g. "abbc", "abbbc", "ddef")

Run `sb-shovel -cmd tidy -conn "cfg|CONNECTION_STRING" -q queuename -pattern "(ab+c|d+ef)"` to verify that all messages get picked up.
```
connecting to CONNECTION_STRING
Tidy executing as a dry run. Pass '-x' to action
[status] identified abbc in message
[status] identified abbbc in message
[status] identified ddef in message
finished processing messages
```
Run `sb-shovel -cmd tidy -conn "cfg|CONNECTION_STRING" -q queuename -pattern "(ab+c|d+ef)" -x` to execute
```
connecting to CONNECTION_STRING
[status] identified abbc in message
[status] identified abbbc  in message
panic: send on closed channel

goroutine 66 [running]:
github.com/aagoldingay/sb-shovel/sbcontroller.(*ServiceBusController).TidyMessages.func1(0xc00018e000)
        C:/.../github.com/aagoldingay/sb-shovel@v0.6.0/sbcontroller/controller.go:336 +0x13e
created by github.com/aagoldingay/sb-shovel/sbcontroller.(*ServiceBusController).TidyMessages.func2
        C:/.../github.com/aagoldingay/sb-shovel@v0.6.0/sbcontroller/controller.go:349 +0xaf
```
This leaves 1 message on the queue

Windows, v0.6.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,13 @@
+# 0.6.1
+
+FIXED
+- `tidy` command `send on closed channel`
+    - Reordered deferred function call stack, to correctly wait for goroutines to complete their operations before cancelling context, which lead to prematurely closing the channel.
+
+CHANGED
+- `delete` command
+    - Reordered deferred function call stack, as in `tidy`.
+
 # 0.6.0
 
 ADDED

--- a/README.md
+++ b/README.md
@@ -4,6 +4,48 @@ sb-shovel is a CLI tool, written in Go, designed to speed up processing Azure Se
 
 This began as a locally developed tool to assist in performing operations on dead-letter queues, in Production environments, with large message volumes. It has proven to be invaluable for myself, so here's hoping somebody else will find this useful!
 
+## Usage
+
+Run one of the following commands to get a complete breakdown of the available commands:
+
+```
+sb-shovel.exe
+sb-shovel.exe -help
+```
+
+Peek-dump queue contents to a local directory:
+
+```
+sb-shovel.exe -cmd pull -conn "<servicebus_connection_string>" -q queueName
+```
+
+Purge the contents of an entire dead-letter queue
+
+```
+sb-shovel.exe -cmd delete -conn "<servicebus_connection_string>" -q queueName -dlq -all
+```
+
+## Installation and Running
+
+Install and set up your Go (1.17+) environment (see main README)
+
+Once done, you can now clone to GOROOT or [GOPATH](https://www.digitalocean.com/community/tutorials/understanding-the-gopath), and set up the environment
+
+Easy: 
+```
+go install github.com/aagoldingay/sb-shovel
+```
+
+Git clone:
+```
+cd $GOPATH/src
+git clone <repo>
+cd sb-shovel
+go mod init
+go mod download
+go run main.go commands.go
+```
+
 ## Project Structure
 
 ```
@@ -56,53 +98,11 @@ sb-shovel
     │   variables.tf
 ```
 
-## Installation and Running
-
-Install and set up your Go (1.17+) environment (see main README)
-
-Once done, you can now clone to GOROOT or [GOPATH](https://www.digitalocean.com/community/tutorials/understanding-the-gopath), and set up the environment
-
-Easy: 
-```
-go install github.com/aagoldingay/sb-shovel
-```
-
-Git clone:
-```
-cd $GOPATH/src
-git clone <repo>
-cd sb-shovel
-go mod init
-go mod download
-go run main.go commands.go
-```
-
 ## Building
 In the same directory as main.go, this is all you need:
 
 ```
 go build
-```
-
-## Usage
-
-Run one of the following commands to get a complete breakdown of the available commands:
-
-```
-sb-shovel.exe
-sb-shovel.exe -help
-```
-
-Peek-dump queue contents to a local directory:
-
-```
-sb-shovel.exe -cmd pull -conn "<servicebus_connection_string>" -q queueName
-```
-
-Purge the contents of an entire dead-letter queue
-
-```
-sb-shovel.exe -cmd delete -conn "<servicebus_connection_string>" -q queueName -dlq -all
 ```
 
 ## Testing

--- a/main.go
+++ b/main.go
@@ -14,7 +14,7 @@ var all, isDlq, delay, help, execute bool
 var maxWriteCache int
 var commandList = map[string]bool{"config": true, "delete": true, "pull": true, "requeue": true, "send": true, "tidy": true}
 
-var version = "v0.6.0"
+var version = "v0.6.1"
 
 func outputCommands() string {
 	s := ""

--- a/sbcontroller/controller.go
+++ b/sbcontroller/controller.go
@@ -93,9 +93,9 @@ func (sb *ServiceBusController) DeleteManyMessages(errChan chan error, total int
 	var wg sync.WaitGroup
 
 	processMessage := func(m *servicebus.Message) {
-		defer wg.Done()
 		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Millisecond)
 		defer cancel()
+		defer wg.Done()
 		m.Complete(ctx)
 	}
 
@@ -111,6 +111,7 @@ func (sb *ServiceBusController) DeleteManyMessages(errChan chan error, total int
 		wg.Add(1)
 		go processMessage(m)
 		if count == total {
+			wg.Wait()
 			cancel()
 		}
 		return nil
@@ -118,7 +119,6 @@ func (sb *ServiceBusController) DeleteManyMessages(errChan chan error, total int
 		errChan <- err
 		return
 	}
-	wg.Wait()
 }
 
 // DisconnectQueues performs both DisconnectSource and DisconnectTarget.
@@ -322,9 +322,9 @@ func (sb *ServiceBusController) TidyMessages(errChan chan error, rex *regexp.Reg
 	var wg sync.WaitGroup
 
 	processMessage := func(m *servicebus.Message) {
-		defer wg.Done()
 		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Millisecond)
 		defer cancel()
+		defer wg.Done()
 
 		result := rex.Find(m.Data)
 
@@ -348,6 +348,7 @@ func (sb *ServiceBusController) TidyMessages(errChan chan error, rex *regexp.Reg
 		wg.Add(1)
 		go processMessage(m)
 		if count == total {
+			wg.Wait()
 			cancel()
 		}
 		return nil
@@ -355,7 +356,6 @@ func (sb *ServiceBusController) TidyMessages(errChan chan error, rex *regexp.Reg
 		errChan <- err
 		return
 	}
-	wg.Wait()
 }
 
 func (sb *ServiceBusController) closeQueue(q *servicebus.Queue) error {

--- a/sbcontroller/controller_integration_test.go
+++ b/sbcontroller/controller_integration_test.go
@@ -467,9 +467,9 @@ func Test_ServiceBusController_Send_And_Delete_Many(t *testing.T) {
 		done = true
 	}
 
-	time.Sleep(2 * time.Second)
-
 	close(eChan)
+
+	time.Sleep(2 * time.Second)
 
 	c, err = sb.GetSourceQueueCount()
 	if c != 0 {
@@ -547,9 +547,9 @@ func Test_ServiceBusController_Send_And_Delete_Trigger_Status(t *testing.T) {
 		done = true
 	}
 
-	time.Sleep(2 * time.Second)
-
 	close(eChan)
+
+	time.Sleep(2 * time.Second)
 
 	c, err = sb.GetSourceQueueCount()
 	if c != 0 {
@@ -591,13 +591,16 @@ func Test_ServiceBusController_TidyMessages_Success(t *testing.T) {
 		t.Error(err)
 	}
 
-	err = sb.SendJsonMessage(false, []byte("abbc"))
-	if err != nil {
-		t.Error(err)
+	for i := 0; i < 2; i++ {
+		err = sb.SendJsonMessage(false, []byte("abbc"))
+		if err != nil {
+			t.Error(err)
+		}
+
 	}
 
 	c, err = sb.GetSourceQueueCount()
-	if c != 1 {
+	if c != 2 {
 		t.Errorf("Unexpected queue count: %d", c)
 	}
 	if err != nil {
@@ -626,7 +629,7 @@ func Test_ServiceBusController_TidyMessages_Success(t *testing.T) {
 	}
 
 	c, err = sb.GetSourceQueueCount()
-	if c != 1 {
+	if c != 2 {
 		t.Errorf("Unexpected queue count: %d", c)
 	}
 	if err != nil {
@@ -648,7 +651,9 @@ func Test_ServiceBusController_TidyMessages_Success(t *testing.T) {
 		done = true
 	}
 
-	time.Sleep(5 * time.Second)
+	close(eChan)
+
+	time.Sleep(2 * time.Second)
 
 	c, err = sb.GetSourceQueueCount()
 	if c != 0 {
@@ -657,9 +662,6 @@ func Test_ServiceBusController_TidyMessages_Success(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-
-	fmt.Println("closing eChan")
-	close(eChan)
 
 	err = sb.DisconnectSource()
 	if err != nil {

--- a/test_files/cmd_tidy_test.txt
+++ b/test_files/cmd_tidy_test.txt
@@ -1,0 +1,3 @@
+abbc
+abbbc
+ddef

